lib/puppet/provider/consul_policy/default.rb

```diff
@@ -0,0 +1,189 @@
+require 'json'
+require 'net/http'
+require 'uri'
+require File.expand_path(File.join(File.dirname(__FILE__), "..", "..", "..", "puppet_x", "consul", "acl_base.rb"))
+
+Puppet::Type.type(:consul_policy).provide(
+    :default
+) do
+  mk_resource_methods
+
+  def self.prefetch(resources)
+    resources.each do |name, resource|
+      rules_encoded = encode_rules(resource[:rules])
+
+      all_policies = list_policies(resource[:acl_api_token], resource[:hostname], resource[:port], resource[:protocol], resource[:api_tries])
+
+      if resource[:id] == ''
+        existing_policy = all_policies.select{|policy| policy.name == name}
+      else
+        existing_policy = all_policies.select{|policy| policy.id == resource[:id]}
+
+        if existing_policy.empty?
+          Puppet.warning("Unable to find any existing Consul ACL policy by specified ID=#{resource[:id]}")
+          resource[:ensure] = :absent
+          return
+        end
+      end
+
+      if existing_policy.length > 0
+        existing_policy = existing_policy.first
+        resource[:id] = existing_policy.id
+
+        existing_policy.rules = @client.get_policy_rules(existing_policy.id, resource[:api_tries])
+      else
+        existing_policy = nil
+      end
+
+      resource.provider = new({}, @client, rules_encoded, existing_policy)
+    end
+  end
+
+  def self.encode_rules(rules)
+    encoded = []
+
+    rules.each do |rule|
+      encoded.push("#{rule['resource']} \"#{rule['segment']}\" {\n  policy = \"#{rule['disposition']}\"\n}")
+    end
+
+    encoded.join("\n\n")
+  end
+
+  def self.list_policies(acl_api_token, hostname, port, protocol, tries)
+    @all_policies ||= nil
+    if @all_policies
+      return @all_policies
+    end
+
+    @client ||= ConsulACLPolicyClient.new(hostname, port, protocol, acl_api_token)
+    @all_policies = @client.get_all_policies(tries)
+    @all_policies
+  end
+
+  def initialize(messages, client = nil, rules_encoded = '', existing_policy = nil)
+    super(messages)
+    @property_flush = {}
+
+    @client = client
+    @rules_encoded = rules_encoded
+    @existing_policy = existing_policy
+  end
+
+  def exists?
+    @property_hash[:ensure] = :present
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def flush
+    if @resource[:ensure] == :absent
+      if @existing_policy
+        @client.delete_policy(@existing_policy.id)
+        Puppet.notice(" Deleted Consul ACL policy #{@existing_policy.name} (ID: #{@existing_policy.id})")
```

Comment:
nit: There is a leading space here in the log line

Reply:
Thx, space has been removed.